### ISSUE TITLE
Long-press should initiate multiselect, not show context menu

### ISSUE THREAD

**Issue by jancborchardt** [owncloud/android#899]
@davivel @rperezb @tobiasKaminsky 

(Edited the title after @joeplus's comment.)

**Comment by joeplus** [comment]
May I make another comment regarding the long press? Since KitKat Google introduced a new usage of the [long press gesture](https://developer.android.com/design/patterns/selection.html) (at the bottom of the page), which now should be used for selection of items. The actions should be moved from the pop up to the action bar. Is this something you guys have already considered to implement long term (maybe with the new drawer and action bar combo)? For now a header makes sense of course!

**Comment by masensio** [comment]
As @joeplus  said, we can introduce the new long press gesture.
![selection_cab_example](https://cloud.githubusercontent.com/assets/3862330/6389174/fc6a2aa6-bda1-11e4-928b-eec082809119.png)

This would mark the selected file and would improve the appearance of the contextual menu, if the pop up is replaced by a  contextual action bar (CAB). (https://developer.android.com/design/patterns/selection.html )
![selection_cab_big](https://cloud.githubusercontent.com/assets/3862330/6389380/d425744a-bda3-11e4-9def-62d13b9fb2f8.png)

In this case, we need to decide which options have to appear in the contextual menu.

**Comment by jancborchardt** [comment]
Yeah – long-press should ideally initiate multi-select mode. As already mentioned in the past: https://github.com/owncloud/android/issues/786#issuecomment-66282140

Then we need to add a 3-dots-icon to each row (on the right) which then invokes the context menu for each file.

**Comment by joeplus** [comment]
Sounds great! [Here](https://keepsafe.github.io/2014/11/19/building-a-custom-overflow-menu.html) is an example of how the three dots could be implemented, just in case.

Another question: Is the content for the listitem popup the exactly the same as the actionbar popup, thus having to different places for the same actions?

**Comment by stefan-niedermann** [comment]
duplicate of #426 ?

**Comment by radziu82** [comment]
No Not duplicate. 426 is closed. This is needed basic functionality.

**Comment by jancborchardt** [comment]
@davivel @masensio @rperezb @MTRichards where are we at with this? Deletion of multiple files is basically impossible (that is, a huge pain) due to multiselect not existing.

**Comment by rperezb** [comment]
@jancborchardt for the moment this is not expected.
Certainly, multiselection is needed as well as the improve on the UI, roadmap

**Comment by jancborchardt** [comment]
Let’s move it up then @MTRichards @rperezb

**Comment by tobiasKaminsky** [comment]
PR will be #1090 by @AndyScherzinger

**Comment by jancborchardt** [comment]
@tobiasKaminsky isn’t that just about multiselect for upload?

**Comment by AndyScherzinger** [comment]
Haven't had this issue in the radar yet

**Comment by tobiasKaminsky** [comment]
Ah. Maybe. I just had read the title. 
And then it would fit, right? ;)
@jancborchardt If it is just for upload, maybe you can clarify it in the issue description?

**Comment by jancborchardt** [comment]
@tobiaskaminsky well this issue is about the app in general, everything _excwpt_ the uploads view. ;) The normal files list, multiselect for that.

**Comment by tobiasKaminsky** [comment]
I would also say that this is the next really big and important improvement that we should challenge right after #819 (which I  have nearly finished).

**Comment by tobiasKaminsky** [comment]
As @masensio said here https://github.com/owncloud/android/issues/899#issuecomment-76148522 a long press would also bring the action for one single file, so "Then we need to add a 3-dots-icon to each row (on the right) which then invokes the context menu for each file." as @jancborchardt said (https://github.com/owncloud/android/issues/899#issuecomment-76180502) should not be needed, right?

**Comment by tobiasKaminsky** [comment]
Currently design with no functionality.
The checkboxes are only shown if one item is selected and disappear if no item is selected.
@jancborchardt the color of the selected items is set to #1D2D44, but seems to be lightened up by android...
(Of course the position of th checkbox in the square images is not correct).

![multi1](https://cloud.githubusercontent.com/assets/5836855/9978219/f6375cb0-5f24-11e5-8e36-bebd92c02894.jpg)
![multi2](https://cloud.githubusercontent.com/assets/5836855/9978218/f6058e42-5f24-11e5-9ca9-21cec6862508.jpg)

**Comment by tobiasKaminsky** [comment]
CAB will include:
if one item selected: all items from currently long press menu
if > 1 item seletced: download, move, copy, delete

**Comment by jancborchardt** [comment]
Make the color a light grey then, that will look much better (so it looks like #f8f8f8 as on the web interface).

Also, the grid lines in thumbnaip view shouldn't be there.

**Comment by AndyScherzinger** [comment]
The grey coloring gets even more important when #1090 is merged, due to the accent colored ticked checkboxes

**Comment by jancborchardt** [comment]
Yeah, also because the boxes themselves are already a very present indication. Only a slight row highlight is needed, if at all.

**Comment by tobiasKaminsky** [comment]
@jancborchardt no grid lines? You mean that the images are without a gap side by side?
This has nothing to do with this issue and also I am not sure if this will look good?
Google's gallery app also has a little space

**Comment by jancborchardt** [comment]
@tobiasKaminsky I mean the grey lines between the folders in your second screenshot (thumbnail view) from https://github.com/owncloud/android/issues/899#issuecomment-141708507 – they don’t look good and are not there by default. ;)

**Comment by tobiasKaminsky** [comment]
Is it okay, that a long click always lead to multi-select?
Maybe it is better to use the K9 way:
Long press opens the current menu as always, but with an additional entry "select" and only this will initiate multi-select

**Comment by tobiasKaminsky** [comment]
@jancborchardt Regarding the grey lines: I will look into it, I suspect that this is a default behaviour of android for multi-select. But I hope that I can disable it ;)

**Comment by jancborchardt** [comment]
@tobiasKaminsky no – long-press should always lead to multi-select. The menu should be brought up by an extra 3-dots icon on the right of the row, just like in the mobile web interface. Ok?

(K9 Mail is honestly not a good inspiration when it comes to interface design. ;)

**Comment by tobiasKaminsky** [comment]
> The menu should be brought up by an extra 3-dots icon on the right of the row, just like in the mobile web interface.

This will not work for grid mode.
So currently I would do it this way, to show in menu (three dots in the CAB [action bar above]):
if one item selected: all items from currently long press menu
if > 1 item selected: only download, move, copy, delete

**Comment by jancborchardt** [comment]
@tobiasKaminsky grid view should have the 3-dots menu too, to the right of the filename / details below the filetype icon, and for images in the bottom right corner.

**Comment by tobiasKaminsky** [comment]
Is this not too much overload? 
The stock google gallery app is it doing the same way as I have described it.

**Comment by jancborchardt** [comment]
Hm, ok – should then the behavior differ between list view and thumbnail view? Might indeed be better to not show an action icon over every thumbnail as it will obstruct the pictures.

**Comment by tobiasKaminsky** [comment]
I would keep the behaviour the same for list and grid view.
Then I will do it this way. 
We can change it later anyway ;)

**Comment by jancborchardt** [comment]
@tobiasKaminsky yeah, good call. As long as it follows nicely designed and established Android apps then it’s good.

**Comment by tobiasKaminsky** [comment]
I have created a Pull Request: #1176, let us discuss everything else there.
Currently the multi select is not working.
But @AndyScherzinger maybe you can have a look from where these grey lines are coming?

**Comment by AndyScherzinger** [comment]
You mean on the grid view? sure!

**Comment by AndyScherzinger** [comment]
Update: fixed the "grid coloring" which actually is the background color, I had to change the setup.xml color definition see https://github.com/owncloud/android/commit/6a8e5930cd7045a5d054bd269a14aea1b7b0d3e4 which is in line with the color changes proposed by @jancborchardt anyways and which I already added to the material branches (buttons, fab). So this won't by a merge problem in the future :)

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger Thanks again :)
Had yesterday big problems, but now it is working :)
PR should be done by end of the week. For my daily use this will be one of the best enhancement after remote thumbnails :)

**Comment by AndyScherzinger** [comment]
Nice! Glad I could help :)

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky I added further icon work for material design purposes :) - One bug I found though is when you select just one them and rotate the screen the whole menu in the action bar is gone. Haven't investigated why this happens though.

**Comment by tobiasKaminsky** [comment]
Can you open a new issue for this?

**Comment by AndyScherzinger** [comment]
Done, please see #1411 I goes it is a save instance issue.

**Comment by malkomich** [comment]
Hi guys,
I have been reviewing the PR #1176 and I have seen that the checkboxes are not always aligned:
![folder](https://cloud.githubusercontent.com/assets/3848202/13701329/831f2988-e787-11e5-8432-59e5e55b4326.png)
![gridview](https://cloud.githubusercontent.com/assets/3848202/13701089/a374c2a8-e785-11e5-997c-017265a5d7fc.png)

In list view they could be aligned right always regardless if there are share icons or not, or maybe just keep the highlight and the top bar information about items selected and not the checkboxes.

For the grid view, the checkboxes could be more neccesary, but they could be shown at the top left size of each item to keep alignment:

About the favorite options, it would be better hide these options on multiple selection when a folder is selected, because it is a feature for files only.

**Comment by AndyScherzinger** [comment]
@malkomich something seems of since the multi select branch implementation has other icons and some newer themeing, see screenshots here: https://github.com/owncloud/android/issues/1434

**Comment by tobiasKaminsky** [comment]
I have updated the branch with latest master.
Now it looks this way:
![multi_list](https://cloud.githubusercontent.com/assets/5836855/13708743/975aa732-e7b0-11e5-9a95-3a61b3543afd.png)
![multi_grid](https://cloud.githubusercontent.com/assets/5836855/13708744/97604930-e7b0-11e5-90d5-eb8245b99569.png)

**Comment by AndyScherzinger** [comment]
I fixed the layouting (and coloring) for the action bar and the list items, see https://github.com/owncloud/android/commit/0ef2389226c97d473e10fa3cd0ad3e0157576a3c
![device-2016-03-12-125532](https://cloud.githubusercontent.com/assets/1315170/13722820/05a4e346-e852-11e5-9583-4c0ea4f53ebb.png)

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky I think @malkomich's point is that due to the fact that some elements are rendered with the file name so the check box is aligned differently. @jancborchardt any idea how this could be solved? The only thing I I can think of is to find a smart way to center the checkbox on each item or we need to find. Google photos for example has the checkbox in the top left corner.

**Comment by tobiasKaminsky** [comment]
For me there are two different approaches, when entering multiselect:
- show only the checkmarks, as the other icons are unimportant
- keep every icon: move checkmarks to top left

**Comment by AndyScherzinger** [comment]
True, not showing that info in multi select mode could also be an option I guess. So waiting for @jancborchardt's input since he is the expert :)

**Comment by tobiasKaminsky** [comment]
Rethinking it, is not such a good idea, I think: what if you want to select all favorite items to unfavorite them?
So we should find a way to have the checkboxes shown with all other items!

**Comment by AndyScherzinger** [comment]
Yeah, that is right. So sticking with top left is probably the easiest approach at the moment.

**Comment by tobiasKaminsky** [comment]
Do you want to do this in the multiSelect branch? :)

**Comment by AndyScherzinger** [comment]
Sure thing :) Gimme a second :grin:

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you take care of the rotation bug regarding the action bar menu? :)

**Comment by tobiasKaminsky** [comment]
I will do this tomorrow morning

**Comment by AndyScherzinger** [comment]
Sweet! :)
Changes for the checkbox have been pushed to the branch.
![device-2016-03-13-135019](https://cloud.githubusercontent.com/assets/1315170/13728768/c0e16da2-e922-11e5-84b8-0aa2a1b83a47.png)

**Comment by tobiasKaminsky** [comment]
:+1: looks great :)

**Comment by AndyScherzinger** [comment]
I now also commited thhttps://github.com/owncloud/android/commit/cca2e60e4932c1351883d358e36322a5f8514baae logic to show hide the fab entering/exiting multi selction mode :grimacing: 

One thing I realized, might be something to deactivate with a flag is that if you do move or copy the file/folder chooser dialog also has multi selection mode which probably doesn't make sense.

**Comment by davivel** [comment]
In, and ready to go with version 2.1.0.
